code/modules/GOAP/GOAPAgent.dm

```diff
@@ -0,0 +1,228 @@
+GLOBAL_LIST_INIT(dangerous_turfs, typecacheof(list(
+	/turf/open/lava,
+	/turf/open/chasm)))
+
+
+#define STATE_IDLE		0
+#define STATE_MOVINGTO	1
+#define STATE_ACTING	2
+
+//#define GOAP_DEBUG_ENABLE "fart"
+
+#ifdef GOAP_DEBUG_ENABLE
+#define goap_debug(string) world.log << "GOAP: [string]"
+#else
+#define goap_debug(string) ;
+#endif
+/datum/goap_agent
+	var/brain_state = STATE_IDLE
+	var/list/our_actions //The actions available to us (/datum/goap_action)
+	var/list/action_queue //The actions of our current plan (if any) (/datum/goap_action)
+
+	//How we plan (Imagine a specific planner that ALWAYS prefers actions involving murder, regardless of cost, bloodlust agent!)
+	var/datum/goap_planner/planner = /datum/goap_planner
+
+	//Provides Info to the agent
+	var/datum/goap_info_provider/info = /datum/goap_info_provider
+
+	var/atom/agent //The actual atom that uses this brain (who knows, maybe you want to give a donut intelligence?)
+
+	var/list/path
+	var/turf/dest
+	var/turf/last_node
+	var/tries = 0
+	var/obj/item/card/id/given_pathfind_access
+	var/movement_type = 1 // 1 = normal A*(expensive but near perfect), 2 = Fake A*(rarely works), 3 = Dumb Movement(cheapest)
+	var/turf/current_loc
+	var/is_megafauna = FALSE
+	var/actions_halted = FALSE
+	var/already_acting = FALSE
+	var/works_when_alone = FALSE
+
+	var/fuck_you_astar = FALSE
+
+/datum/goap_agent/proc/has_action(var/ACT)
+	for(var/datum/goap_action/A in our_actions)
+		if(istype(A, ACT) && !A.OnCooldown())
+			return TRUE
+	return FALSE
+
+/datum/goap_agent/New()
+	our_actions = list()
+	action_queue = list()
+
+	if(ispath(info))
+		info = new info()
+	else
+		goap_debug("OH GOD HELP ME I DONT UNDERSTAND THE WORLD")
+		return
+
+	if(ispath(planner))
+		planner = new planner()
+	else
+		goap_debug("OH GOD HELP ME I DONT KNOW HOW TO THINK STRAIGHT")
+		return
+
+	START_PROCESSING(SSgoap, src)
+
+/datum/goap_agent/Destroy()
+	STOP_PROCESSING(SSgoap, src)
+	return ..()
+
+/datum/goap_agent/proc/able_to_run()
+	if(!agent)
+		STOP_PROCESSING(SSgoap, src)
+		qdel(src)
+		return FALSE
+	if(works_when_alone)
+		return TRUE
+	for(var/I in GLOB.mob_list)
+		var/mob/M = I
+		if(M != null && M.z == agent.z && M.client && istype(M, /mob/living) && get_dist(M, agent) <= 14)
```

Comment:
use `var/I in GLOB.living_mob_list & GLOB.alive_mob_list & SSmob.zlevel_mobs[agent.z]` and you can eliminate all but the last check

The name of that last var isn't right though, @vuonojenmustaturska knows the right one